### PR TITLE
Sitewide Sales Version 1.4 Release

### PR DESCRIPTION
**Pull request by ideadude**
Merging dev into master for the new release. Will review and add notes.

### CODE REVIEW

File: classes/class-swsales-about.php [hunk 1]
```diff
@@ -38,25 +38,12 @@ public static function show_about_page() { ?>
 						'strong' => array(),
 						'em' => array(),		);
 				?>
-				<p><?php echo sprintf( wp_kses( __( 'Sitewide Sales helps you run Black Friday, Cyber Monday, or other flash sales on your WordPress-powered eCommerce or membership site. We currently offer integration for <a href="%s" title="Paid Memberships Pro" target="_blank">Paid Memberships Pro</a>, <a href="%s" title="WooCommerce" target="_blank">WooCommerce</a>, and <a href="%s" title="Easy Digital Downloads" target="_blank">Easy Digital Downloads</a>.', 'sitewide-sales' ), $allowed_html ), 'https://www.paidmembershipspro.com/?utm_source=sitewide-sales&utm_medium=about&utm_campaign=homepage', 'https://woocommerce.com', 'https://easydigitaldownloads.com' ); ?></p>
+				<p><?php echo wp_kses( __( 'Sitewide Sales helps you run Black Friday, Cyber Monday, or other flash sales on your WordPress-powered eCommerce or membership site.', 'sitewide-sales' ), $allowed_html ); ?></p>
```
Comment (ideadude) [inline classes/class-swsales-about.php:41]:
```suggestion
				<p><?php esc_html_e( 'Sitewide Sales helps you run Black Friday, Cyber Monday, or other flash sales on your WordPress-powered eCommerce or membership site.', 'sitewide-sales' ); ?></p>
```
Doesn't need wp_kses since it doesn't have links in it anymore.

File: classes/class-swsales-about.php [hunk 2]
```diff
@@ -38,25 +38,12 @@ public static function show_about_page() { ?>
...
-					<li><?php esc_html_e( 'If you\'re running a sale on membership, which existing members (if any) do you want to know about the sale?', 'sitewide-sales' ); ?></li>
-					<li><?php esc_html_e( 'What general look and feel do you want to use as part of the marketing surrounding your sale?', 'sitewide-sales' ); ?></li>
-				</ul>
-
-				<h2><?php esc_html_e( "We're here to help make your sale a measurable success.", 'sitewide-sales' ); ?></h2>
-
-				<?php
-					echo '<p>' . wp_kses( __( 'Check out the Sitewide Sales documentation site for additional setup instructions, sample landing page and banner content, as well as developer documentation to further extend the templates, reporting, and integration options.', 'sitewide-sales' ), $allowed_html ) . '</p>';
-				?>
+				<p><?php echo wp_kses( __( 'Check out the Sitewide Sales documentation site for additional setup instructions, sample landing page and banner content, as well as developer documentation to further extend the templates, reporting, and integration options.', 'sitewide-sales' ), $allowed_html ); ?></p>
```
Comment (ideadude) [inline classes/class-swsales-about.php:46]:
```suggestion
				<p><?php esc_html_e( 'Check out the Sitewide Sales documentation site for additional setup instructions, sample landing page and banner content, as well as developer documentation to further extend the templates, reporting, and integration options.', 'sitewide-sales' ); ?></p>
```

File: classes/class-swsales-metaboxes.php [hunk 3]
```diff
@@ -313,6 +313,26 @@ public static function display_step_type( $post ) {
 				// Filter to add custom settings from module .
 				do_action( 'swsales_after_choose_sale_type', $cur_sale );
 				?>
+				<tr>
+					<th><label for="swsales_hide_for_roles"><?php esc_html_e( 'Hide Sale by Role', 'sitewide-sales' ); ?></label></th>
```
Comment (ideadude) [inline classes/class-swsales-metaboxes.php:317]: for attributes need to point to an id, not a name attribute. So probably change to swsales_hide_for_roles_select

File: classes/class-swsales-metaboxes.php [hunk 4]
```diff
@@ -409,26 +429,33 @@ public static function display_step_banner( $post ) {
...
-					<th><label for="swsales_hide_banner_by_role"><?php esc_html_e( 'Hide Banner by Role', 'sitewide-sales' ); ?></label></th>
-					<td>
-						<input type="hidden" name="swsales_hide_banner_by_role_exists" value="1" />
-						<select multiple class="swsales_option" id="swsales_hide_banner_by_role_select" name="swsales_hide_banner_by_role[]">
+				<?php
+					// Old filter to hide banner by role. This is being deprecated in place of the field swsales_hide_for_roles.
+					$hide_for_roles = json_decode( $cur_sale->get_meta_value( 'swsales_hide_banner_by_role', '[]' ) );
+					if ( ! empty( $hide_for_roles ) ) { ?>
+						<tr>
+							<th><label for="swsales_hide_banner_by_role"><?php esc_html_e( 'Hide Banner by Role', 'sitewide-sales' ); ?></label></th>
```
Comment (ideadude) [inline classes/class-swsales-metaboxes.php:437]: for needs to point to an id, so probably change to swsales_hide_banner_by_role_select

Comment (ideadude) [reply classes/class-swsales-metaboxes.php:437]: A way to test is clicking on the label should select the associated input/field.

File: classes/class-swsales-metaboxes.php [hunk 5]
```diff
@@ -707,6 +734,20 @@ public static function save_swsales_metaboxes( $post_id, $post ) {
 			update_post_meta( $post_id, 'swsales_sale_type', sanitize_text_field( $_POST['swsales_sale_type'] ) );
 		}
 
+		if ( ! empty( $_POST['swsales_hide_banner_by_role'] ) && is_array( $_POST['swsales_hide_banner_by_role'] )) {
+			$swsales_hide_banner_by_role = array_map( 'sanitize_text_field', $_POST['swsales_hide_banner_by_role'] );
+			update_post_meta( $post_id, 'swsales_hide_banner_by_role', wp_json_encode( $swsales_hide_banner_by_role ) );
```
Comment (ideadude) [inline classes/class-swsales-metaboxes.php:739]: Note here that these wp_json_encode calls are not really needed as far as I know (assuming the wp_json_decode calls are also removed when the post meta is gotten), because the post meta get/set functions will serialize/deserialize the array anyway.

The serialize functions (automatically applied to arrays get/set in meta) are like old school json encode/decode.

But it seems like we were already saving these values this way, so let's continue to do that... better than coding some way to migrate old meta or account for both formats.

If I'm mistaken and the json encode is doing something important here, let us know. Update the comments. And share with the dev team.

File: classes/class-swsales-reports.php [hunk 6]
```diff
@@ -15,6 +15,7 @@ public static function init() {
 		add_action( 'wp_enqueue_scripts', array( __CLASS__, 'enqueue_tracking_js' ) );
 		add_action( 'wp_ajax_swsales_ajax_tracking', array( __CLASS__, 'ajax_tracking' ) );
 		add_action( 'wp_ajax_nopriv_swsales_ajax_tracking', array( __CLASS__, 'ajax_tracking' ) );
+		add_action("wp_ajax_sws_stats_csv", array(__CLASS__, "sws_stats_csv"));
```
Comment (ideadude) [inline classes/class-swsales-reports.php:18]: Are we pretty consistent about using swsales_ as the prefix?

If so, let's update the AJAX call and related callback function names to swsales_...

File: classes/class-swsales-reports.php [hunk 7]
```diff
@@ -28,6 +29,36 @@ public static function add_reports_page() {
 		);
 	}
 
+	/**
+	 * Handles the Sales Export
+	 */
+	public static function sws_stats_csv() {
+		require_once(dirname(__FILE__) . "/../adminpages/report-csv.php");
```
Comment (ideadude) [inline classes/class-swsales-reports.php:36]: Can we use one of these constants instead of dirname here?

https://github.com/strangerstudios/sitewide-sales/blob/dev/sitewide-sales.php#L19-L22

File: classes/class-swsales-reports.php [hunk 8]
```diff
@@ -28,6 +29,36 @@ public static function add_reports_page() {
 		);
 	}
 
+	/**
+	 * Handles the Sales Export
+	 */
+	public static function sws_stats_csv() {
```
Comment (ideadude) [inline classes/class-swsales-reports.php:35]: Actually doesn't need the prefix since this is a method in a class.

File: classes/class-swsales-reports.php [hunk 9]
```diff
@@ -42,141 +73,263 @@ public static function show_reports_page() { ?>
 				);
 
 				// Choose sale to show.
-				$sale_to_show = null;
+				$sales_to_show = array();
 				if ( isset( $_REQUEST['sitewide_sale'] ) ) {
-					$sale_to_show = SWSales_Sitewide_Sale::get_sitewide_sale( $_REQUEST['sitewide_sale'] );
+					if ( ! is_array( $_REQUEST['sitewide_sale'] ) ) {
+						$_REQUEST['sitewide_sale'] = array( $_REQUEST['sitewide_sale'] );
```
Comment (ideadude) [inline classes/class-swsales-reports.php:79]:
```suggestion
						$_REQUEST['sitewide_sale'] = array( $_REQUEST['sitewide_sale'] ); // Sanitized below.
```

File: classes/class-swsales-reports.php [hunk 10]
```diff
@@ -42,141 +73,263 @@ public static function show_reports_page() { ?>
...
-								?>
-								<option value="<?php esc_attr_e( $sale->get_id() ); ?>" <?php echo( esc_html( $selected_modifier ) ); ?>>
-									<?php echo( esc_html( $sale->get_name() ) ); ?>
-								</option>
+						<div class="swsales_reports-filters">
+							<?php 
+
+							for ($i = 0; $i < 2; $i++) {
+								$class_modifier = ( $i === 0 ) ? 'left' : 'right';
+								$label_wording = ( $i === 0 ) ? 'Show reports for' : 'vs.';
```
Comment (ideadude) [inline classes/class-swsales-reports.php:102]: Need to localize?

File: classes/class-swsales-reports.php [hunk 11]
```diff
@@ -42,141 +73,263 @@ public static function show_reports_page() { ?>
...
+								
+								$selected =  $sales_to_show !== null && array_key_exists( $i, $sales_to_show )  ? $sales_to_show[$i] : null;
+							?>
+								<label for="swsales_select_report_<?php echo $class_modifier ?>"><?php esc_html_e( $label_wording, 'sitewide-sales' ); ?></label>
```
Comment (ideadude) [inline classes/class-swsales-reports.php:106]: Don't use the localization functions (_e or esc_html_e) unless the first parameter is a "literal string", i.e. not a variable.

Comment (ideadude) [inline classes/class-swsales-reports.php:106]:
```suggestion
								<label for="swsales_select_report_<?php echo esc_attr( $class_modifier ); ?>"><?php esc_html_e( $label_wording, 'sitewide-sales' ); ?></label>
```

File: classes/class-swsales-reports.php [hunk 12]
```diff
@@ -42,141 +73,263 @@ public static function show_reports_page() { ?>
...
+								<select id="swsales_select_report_<?php echo $class_modifier ?>"  name="sitewide_sale[]">
```
Comment (ideadude) [inline classes/class-swsales-reports.php:107]:
```suggestion
								<select id="swsales_select_report_<?php echo esc_attr( $class_modifier );?>"  name="sitewide_sale[]">
```

File: classes/class-swsales-reports.php [hunk 13]
```diff
@@ -42,141 +73,263 @@ public static function show_reports_page() { ?>
...
+									<?php
+									if($i == 1) {
+									?>
+									<option selected value="0"><?php esc_html_e( '- Choose One -', 'sitewide-sales' ); ?></option>
+									<?php
+									}
+									foreach ( $all_sitewide_sales as $sitewide_sale_id ) {
+										$sale   = SWSales_Sitewide_Sale::get_sitewide_sale( $sitewide_sale_id );
+										$selected_modifier =  $selected != null && $sale === $selected ? 'selected="selected"' : '';
```
Comment (ideadude) [inline classes/class-swsales-reports.php:116]: Probably not worth changing, but could consider making sure $selected is never null (initialize as '') and using the WP function inline in the option tag below.

https://developer.wordpress.org/reference/functions/selected/

File: classes/class-swsales-reports.php [hunk 14]
```diff
@@ -42,141 +73,263 @@ public static function show_reports_page() { ?>
...
+	 * Show report content for selected Sitewide Sales objects.
 	 *
-	 * @param SWSales_Sitewide_Sale $sitewide_sale to show report for.
+	 * @param Array An array of SWSales_Sitewide_Sale objects to show report for.
+	 * @since TBD.
 	 */
-	public static function show_report( $sitewide_sale ) {
-		if ( ! is_a( $sitewide_sale, 'Sitewide_Sales\classes\SWSales_Sitewide_Sale' ) ) {
+	public static function show_report( $sitewide_sales ) {
+		// Bail if the array comes empty
```
Comment (ideadude) [inline classes/class-swsales-reports.php:158]: Is there any chance people or other code was calling this show_report button and passing in a single report?

If so, we can account for that by check if ! is_array and casting $sitewide_sales to an array if not.

But maybe not worth the extra code if we feel no one would have called this themselves.

File: classes/class-swsales-reports.php [hunk 15]
```diff
@@ -42,141 +73,263 @@ public static function show_reports_page() { ?>
...
+					?>
+					<tr>
+						<th>
+							<div class="swsales_reports-sale-value">
+								<a href="<?php echo esc_url( add_query_arg( array( 'post' => $sitewide_sale->get_id(), 'action' => 'edit' ), admin_url( 'post.php' ) ) ); ?>"><?php echo esc_html( $sitewide_sale->get_name() ); ?></a>
+							</div>
+							<div class="swsales_reports-sale-value-description">
+								<?php
+									printf(
+										wp_kses_post( '%s to %s', 'sitewide-sales' ),
```
Comment (ideadude) [inline classes/class-swsales-reports.php:200]: Is this wp_kses_post lazy? I think we had one or more before around this code. Are we okay with ANY HTML in these values?

File: classes/class-swsales-reports.php [hunk 16]
```diff
@@ -28,6 +29,36 @@ public static function add_reports_page() {
...
+			return;
+		}
+		$csv_export_link = add_query_arg(
+			array(
+				'action' => 'sws_stats_csv',
+				'sitewide_sale' => $sitewide_sale->get_id(),
+			),
+			admin_url( 'admin-ajax.php' ) );
+
+		return esc_url( $csv_export_link );
```
Comment (ideadude) [inline classes/class-swsales-reports.php:59]: Probably better to not escape and return it raw and expect the code calling this to escape.

Okay to leave the esc_url here and ALSO esc_url when the URL is output. Running esc_url twice on the same URL should not break things.

File: classes/class-swsales-reports.php [hunk 17]
```diff
@@ -42,141 +73,263 @@ public static function show_reports_page() { ?>
...
+
+									$start_date = date_create($start_date);
+									$end_date = date_create($end_date);
+
+									// Calculate the date difference
+									$interval = $start_date->diff($end_date);
+									echo esc_html( sprintf( _n( '(%s Day)', '(%s Days)', $interval->days, 'sitewide-sales' ), number_format_i18n( $interval->days ) ) );
+								?>
+							</div>
+							<a target="_blank" href="<?php echo SWSales_Reports::build_CSV_report_link( $sitewide_sale ); ?>" class="button button-primary button-small"><?php esc_html_e( 'Export to CSV', 'sitewide-sales' ); ?></a>
```
Comment (ideadude) [inline classes/class-swsales-reports.php:217]:
```suggestion
							<a target="_blank" href="<?php echo esc_url( SWSales_Reports::build_CSV_report_link( $sitewide_sale ) ); ?>" class="button button-primary button-small"><?php esc_html_e( 'Export to CSV', 'sitewide-sales' ); ?></a>
```

Comment (ideadude) [reply classes/class-swsales-reports.php:217]: Even though we esc_url in the method there, good to esc_url "as late as possible before outputting" again here to be clear to others reading the code.

File: classes/class-swsales-reports.php [hunk 18]
```diff
@@ -366,4 +631,64 @@ public static function ajax_tracking() {
...
+
+	/**
+	 * Given a rate and a boolean indicating if it's a growth or decline, returns the markup representing the rate.
+	 *
+	 * @param Obj can be a float or a String $rate The rate to represent.
+	 * @param Boolean $is_growth Indicates if the rate is a growth or decline.
+	 * @return String with the markup representing the rate.
+	 * @since TBD.
+	 */
+	private static function build_rate_markup( $rate, $is_growth ) {
```
Comment (ideadude) [inline classes/class-swsales-reports.php:686]: Needs escaping.

Be sure to note in a comment when these are output later that they were escaped. Maybe even change the var names later to ..._escaped.

File: classes/class-swsales-reports.php (hunk 8 above)
Comment (kimcoleman) [reply classes/class-swsales-reports.php:35]: So we should rename to just stats_csv?

File: classes/class-swsales-reports.php (hunk 16 above)
Comment (kimcoleman) [reply classes/class-swsales-reports.php]: I removed the esc_url here and only escaping on output now.

File: classes/class-swsales-reports.php (hunk 15 above)
Comment (kimcoleman) [reply classes/class-swsales-reports.php:200]: I don't think we want any HTML actually. I can switch this to just esc_html?

File: classes/class-swsales-reports.php (hunk 18 above)
Comment (kimcoleman) [reply classes/class-swsales-reports.php:686]: Did escaping just in the build_rate_markup() method and added comments

File: classes/class-swsales-reports.php (hunk 11 above)
Comment (kimcoleman) [reply classes/class-swsales-reports.php]: Fixed.

File: classes/class-swsales-reports.php (hunk 10 above)
Comment (kimcoleman) [reply classes/class-swsales-reports.php]: Done :-)

File: classes/class-swsales-reports.php (hunk 7 above)
Comment (kimcoleman) [reply classes/class-swsales-reports.php]: Yep - require_once( SWSALES_DIR . "/adminpages/report-csv.php"); done.

File: classes/class-swsales-reports.php (hunk 6 above)
Comment (kimcoleman) [reply classes/class-swsales-reports.php]: Removed prefix in class method; updated prefix for ajax action.

File: classes/class-swsales-reports.php (hunk 8 above)
Comment (kimcoleman) [reply classes/class-swsales-reports.php]: Removed prefix in class method; updated prefix for ajax action.

File: classes/class-swsales-metaboxes.php (hunk 5 above)
Comment (kimcoleman) [reply classes/class-swsales-metaboxes.php:739]: I asked ChatGPT and seems to think that meta isn't serialized automatically. IDK. This was how we had the level info stored so used the same model for the role info. I don't know if its needed but can test it out.

File: classes/class-swsales-reports.php (hunk 13 above)
Comment (kimcoleman) [reply classes/class-swsales-reports.php:117]: I worked on this a bit but its not liking anyting I try. Not changing for now.
